src/bluesearch/k8s/ner.py

```diff
@@ -0,0 +1,204 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <https://www.gnu.org/licenses/>.
+"""Perform Name Entity Recognition (NER) on a paragraph."""
+import logging
+import os
+from typing import Any
+
+import elasticsearch
+import numpy as np
+import requests
+import tqdm
+from dotenv import load_dotenv
+from elasticsearch.helpers import scan
+
+load_dotenv()
+
+logger = logging.getLogger(__name__)
+
+
+def run(
+    client: elasticsearch.Elasticsearch,
+    index: str = "paragraphs",
+    ner_method: str = "both",
+    force: bool = False,
+) -> None:
+    """Runs the NER pipeline on the paragraphs in the database.
+
+    Parameters
+    ----------
+    client
+        Elasticsearch client.
+    index
+        Name of the ES index.
+    ner_method
+        Method to use to perform NER.
+    """
+
+    # get paragraphs without NER unless force is True
+    if force:
+        query: dict[str, Any] = {"match_all": {}}
+    else:
+        query = {"bool": {"must_not": {"exists": {"field": "ner"}}}}
+    paragraph_count = client.count(index=index, query=query)["count"]
+    logger.info("There are {paragraph_count} paragraphs without embeddings")
+
+    # creates NER for all the documents
+    progress = tqdm.tqdm(
+        total=paragraph_count,
+        position=0,
+        unit=" Paragraphs",
+        desc="Updating NER",
+    )
+    for hit in scan(client, query={"query": query}, index=index):
+        if ner_method == "both":
+            results_ml = run_ml_ner(
+                hit["_source"]["text"],
+                os.environ["BENTOML_NER_ML_URL"]
+            )
+            results_ruller = run_ruler_ner(
+                hit["_source"]["text"],
+                os.environ["BENTOML_NER_RULER_URL"]
+            )
+
+            client.update(index=index, doc={"ner_ml": results_ml}, id=hit["_id"])
+            client.update(index=index, doc={"ner_ruler": results_ruller}, id=hit["_id"])
+
+        elif ner_method == "ml":
+            results = run_ml_ner(
+                hit["_source"]["text"],
+                os.environ["BENTOML_NER_RULER_URL"]
+            )
+            client.update(index=index, doc={"ner_ml": results}, id=hit["_id"])
+        elif ner_method == "ruler":
+            results = run_ruler_ner(
+                hit["_source"]["text"],
+                os.environ["BENTOML_NER_RULER_URL"]
+            )
+            client.update(index=index, doc={"ner_ruler": results}, id=hit["_id"])
+        else:
+            raise ValueError(f"Unknown NER method: {ner_method}")
+
+        progress.update(1)
+        logger.info(f"Updated NER for paragraph {hit['_id']}, progress: {progress.n}")
+
+    progress.close()
+
+
+def run_ml_ner(text: str, url: str) -> list[dict]:
```

Comment:
I am wondering now that the URL is a parameter of the function. Are there any difference between `run_ml_ner` and `run_ml_ruler` ? Maybe we can merge those two functions. What do you think ?